xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-notifiers/xwiki-platform-notifications-notifiers-default/src/main/java/org/xwiki/notifications/notifiers/internal/UserEventDispatcher.java

```diff
@@ -119,81 +114,10 @@ public class UserEventDispatcher implements Disposable
     @Inject
     private RemoteObservationManagerConfiguration remoteObservation;
 
-    @Inject
-    private ExecutionContextManager contextManager;
-
-    @Inject
-    private Execution execution;
-
     @Inject
     private Logger logger;
 
-    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
-
-    private volatile boolean running;
-
-    /**
-     * Start the scheduler.
-     */
-    public void initialize()
-    {
-        // Schedule a job to regularly check if any event prefiltering was missed or a previous run failed
-        this.scheduler.scheduleWithFixedDelay(this::run, 0, 1, TimeUnit.HOURS);
-    }
-
-    /**
-     * Indicate an event just been created.
-     * @param event the event that has been created
-     */
-    public void onEvent(Event event)
-    {
-        try {
-            if (getSupportedEventTypes().contains(event.getType()) && !this.running) {
-                // Make sure to wakeup the dispatcher
-                this.scheduler.execute(this::run);
-            }
-        } catch (EventStreamException e) {
-            this.logger.error("Failed to get supported event types", e);
-        }
-    }
-
-    @Override
-    public void dispose() throws ComponentLifecycleException
-    {
-        // Stop the scheduling
-        this.scheduler.shutdownNow();
-    }
-
-    private void run()
-    {
-        // Indicate the pre-filtering is running
-        this.running = true;
-
-        Thread currenthread = Thread.currentThread();
-
-        // Reduce the priority of this thread since it's not a critical task and it might be expensive
-        currenthread.setPriority(Thread.NORM_PRIORITY - 1);
-        // Set a more explicit thread name
-        currenthread.setName("User event dispatcher thread");
-
-        try {
-            // Initialize a new context for the run
-            this.contextManager.initialize(new ExecutionContext());
-
-            flush();
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-        } catch (Throwable e) {
-            // Catching Throwable to make sure we don't kill the scheduler which triggered this run
-            this.logger.error("Failed to pre-filter events", e);
-        } finally {
-            // Indicate the pre-filtering is not running anymore
-            this.running = false;
-
-            // Remove any remaining context
-            this.execution.removeContext();
-        }
-    }
+    private CompletableFuture<?> lastSave;
```

Comment:
I'm not a big fan of this, it would make much more sense in terms of design to have prefilterEvent return the CompletableFuture to join.